In [51]:
import pandas as pd
import numpy as np

In [28]:
df_ratings = pd.read_excel('Metric-SpreadSheet-Assignment.xlsx', skiprows=1, index_col=0, nrows=10).iloc[:, :10]
df_ratings

11: Star Wars: Episode IV - A New Hope (1977)  12: Finding Nemo (2003)  \
5136                                            4.5                      5.0   
918                                             5.0                      5.0   
2824                                            4.5                      NaN   
860                                             4.0                      4.0   
442                                             3.0                      NaN   
3556                                            4.0                      4.0   
5261                                            NaN                      NaN   
2492                                            4.5                      3.5   
4942                                            NaN                      3.5   
2267                                            NaN                      NaN   

      13: Forrest Gump (1994)  14: American Beauty (1999)  \
5136                      5.0                         4.0   
918                       4.5                         NaN   
2824                      5.0                         NaN   
860                       4.5                         NaN   
442                       3.0                         4.0   
3556                      4.0                         4.0   
5261                      5.0                         NaN   
2492                      3.5                         3.5   
4942                      4.0                         4.0   
2267                      3.5                         NaN   

      22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)  \
5136                                                5.0                   
918                                                 3.0                   
2824                                                4.5                   
860                                                 2.5                   
442                                                 1.0                   
3556                                                3.0                   
5261                                                1.5                   
2492                                                4.0                   
4942                                                3.5                   
2267                                                NaN                   

      24: Kill Bill: Vol. 1 (2003)  \
5136                           5.0   
918                            NaN   
2824                           4.0   
860                            3.0   
442                            5.0   
3556                           5.0   
5261                           0.5   
2492                           4.0   
4942                           4.0   
2267                           5.0   

      38: Eternal Sunshine of the Spotless Mind (2004)  \
5136                                               5.0   
918                                                5.0   
2824                                               NaN   
860                                                NaN   
442                                                3.0   
3556                                               NaN   
5261                                               NaN   
2492                                               NaN   
4942                                               5.0   
2267                                               1.5   

      63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)  77: Memento (2000)  \
5136                                            3.0                 NaN   
918                                             NaN                 5.0   
2824                                            NaN                 5.0   
860                                             4.0                 4.5   
442                                             2.0                 4.0   
3556                                            NaN                 NaN   
5261                                        

In [29]:
df_preds = pd.read_excel('Metric-SpreadSheet-Assignment.xlsx', skiprows=15, index_col=0, nrows=10).iloc[:, :10]
df_preds.columns = df_ratings.columns
df_preds

11: Star Wars: Episode IV - A New Hope (1977)  12: Finding Nemo (2003)  \
Users                                                                           
5136                                             4.1                      4.3   
918                                              4.0                      2.3   
2824                                             4.7                      NaN   
860                                              4.0                      3.0   
442                                              3.6                      NaN   
3556                                             4.3                      4.7   
5261                                             NaN                      NaN   
2492                                             4.1                      2.7   
4942                                             NaN                      2.4   
2267                                             NaN                      NaN   

       13: Forrest Gump (1994)  14: American Beauty (1999)  \
Users                                                        
5136                       4.7                         3.7   
918                        4.6                         NaN   
2824                       4.8                         NaN   
860                        3.8                         NaN   
442                        4.2                         3.0   
3556                       3.7                         2.0   
5261                       4.6                         NaN   
2492                       4.2                         4.3   
4942                       3.3                         5.0   
2267                       3.9                         NaN   

       22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)  \
Users                                                                      
5136                                                 3.8                   
918                                                  2.6                   
2824                                                 3.9                   
860                                                  4.1                   
442                                                  2.2                   
3556                                                 4.4                   
5261                                                 3.5                   
2492                                                 3.3                   
4942                                                 4.2                   
2267                                                 NaN                   

       24: Kill Bill: Vol. 1 (2003)  \
Users                                 
5136                            4.8   
918                             NaN   
2824                            4.2   
860                             3.6   
442                             4.2   
3556                            4.1   
5261                            1.7   
2492                            3.3   
4942                            4.6   
2267                            4.4   

       38: Eternal Sunshine of the Spotless Mind (2004)  \
Users                                                     
5136                                                3.2   
918                                                 4.4   
2824                                                NaN   
860                                                 NaN   
442                                                 2.2   
3556                                                NaN   
5261                                                NaN   
2492                                                NaN   
4942                                                4.4   
2267                                                2.6   

       63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)  77: Memento (2000)  \
Users                                                                      
5136                                          

In [33]:
df_abs_error = (df_preds - df_ratings).abs()
df_abs_error

11: Star Wars: Episode IV - A New Hope (1977)  12: Finding Nemo (2003)  \
Users                                                                           
5136                                             0.4                      0.7   
918                                              1.0                      2.7   
2824                                             0.2                      NaN   
860                                              0.0                      1.0   
442                                              0.6                      NaN   
3556                                             0.3                      0.7   
5261                                             NaN                      NaN   
2492                                             0.4                      0.8   
4942                                             NaN                      1.1   
2267                                             NaN                      NaN   

       13: Forrest Gump (1994)  14: American Beauty (1999)  \
Users                                                        
5136                       0.3                         0.3   
918                        0.1                         NaN   
2824                       0.2                         NaN   
860                        0.7                         NaN   
442                        1.2                         1.0   
3556                       0.3                         2.0   
5261                       0.4                         NaN   
2492                       0.7                         0.8   
4942                       0.7                         1.0   
2267                       0.4                         NaN   

       22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)  \
Users                                                                      
5136                                                 1.2                   
918                                                  0.4                   
2824                                                 0.6                   
860                                                  1.6                   
442                                                  1.2                   
3556                                                 1.4                   
5261                                                 2.0                   
2492                                                 0.7                   
4942                                                 0.7                   
2267                                                 NaN                   

       24: Kill Bill: Vol. 1 (2003)  \
Users                                 
5136                            0.2   
918                             NaN   
2824                            0.2   
860                             0.6   
442                             0.8   
3556                            0.9   
5261                            1.2   
2492                            0.7   
4942                            0.6   
2267                            0.6   

       38: Eternal Sunshine of the Spotless Mind (2004)  \
Users                                                     
5136                                                1.8   
918                                                 0.6   
2824                                                NaN   
860                                                 NaN   
442                                                 0.8   
3556                                                NaN   
5261                                                NaN   
2492                                                NaN   
4942                                                0.6   
2267                                                1.1   

       63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)  77: Memento (2000)  \
Users                                                                      
5136                                          

### Mean Absolute Error (MAE)

1. Compute the mean absolute error for each user.  For example, user 5261 has a mean absolute error of 0.82 (mostly because of a really bad prediction on the Pirates of the Carribean movie).  
1.1  which users have the highest and lowest MAE, and what are those MAE values?
2. Also compute the mean absolute error associated with the predictions for each movie.  For example, Forrest Gump has an MAE of 0.5.  
2.1 which movies have the highest and lowest MAE, and what are those MAE values?

In [107]:
df_abs_error.mean(axis=1).round(3).sort_values(ascending=False)

Users
2267    0.900
3556    0.843
5261    0.820
918     0.800
860     0.786
4942    0.762
5136    0.733
442     0.678
2492    0.589
2824    0.400
dtype: float64

In [108]:
df_abs_error.mean(axis=0).round(3).sort_values(ascending=False)

12: Finding Nemo (2003)                                                               1.167
22: Pirates of the Caribbean: The Curse of the Black Pearl (2003)                     1.089
14: American Beauty (1999)                                                            1.020
38: Eternal Sunshine of the Spotless Mind (2004)                                      0.980
77: Memento (2000)                                                                    0.662
24: Kill Bill: Vol. 1 (2003)                                                          0.644
13: Forrest Gump (1994)                                                               0.500
63: Twelve Monkeys (a.k.a. 12 Monkeys) (1995)                                         0.460
11: Star Wars: Episode IV - A New Hope (1977)                                         0.414
85: Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)    0.380
dtype: float64

3.  Compute the overall MAE three ways (three decimal places is enough):  
3.1 the normal MAE, which averages over all predictions  
3.2 the by-user MAE, which averages the MAEs of each user  
3.3 the by-movie MAE, which averages the MAEs of each movie   

In [63]:
pd.Series(df_abs_error.values.flatten()).dropna().mean().round(3)

np.float64(0.725)

In [62]:
# users
df_abs_error.mean(axis=1).mean().round(3)

np.float64(0.731)

In [64]:
# movies
df_abs_error.mean(axis=0).mean().round(3)

np.float64(0.732)

4.  Next, create a matrix with the squared errors (instead of absolute errors), and compute the MSE and RMSE (root-mean-squared-error) for each user.  Excel and other spreadsheets have a built-in function SQRT() to compute square roots. Be sure that you are computing the MSE first, then taking the square root (if you take the square root of each error before averaging it, you end up with MAE again).  For example, user 5136 has an MSE of 0.7911 and an RMSE of .8894.  
4.1 which users have the highest and lowest RMSE, and what are those RMSE values?


In [81]:
mse_user = (df_preds - df_ratings).pow(2).mean(axis=1)
mse_user.sort_values()

Users
2824    0.224000
2492    0.412222
4942    0.613750
442     0.641111
5136    0.791111
860     0.858571
2267    0.995000
3556    1.075714
5261    1.170000
918     1.470000
dtype: float64

In [109]:
rmse_user = mse_user.pow(.5)
rmse_user.sort_values(ascending=False)

Users
918     1.212436
5261    1.081665
3556    1.037166
2267    0.997497
860     0.926591
5136    0.889444
442     0.800694
4942    0.783422
2492    0.642045
2824    0.473286
dtype: float64

5.  Then, compute the overall MSE and RMSE across all predictions (just the normal way, no need to compute by-user or by-movie).  Again, remember to average all the squared errors to compute MSE, then take the square root of the MSE to get RMSE.

In [94]:
# users
mse = pd.Series((df_preds - df_ratings).pow(2).values.flatten()).dropna().mean()

rmse = np.pow(mse, .5)

print(f'MSE: {mse:.3f}\nRMSE: {rmse:.3f}')

MSE: 0.794
RMSE: 0.891


6.  Next, compute the correlation between ratings and predictions using the built-in CORREL() function.  First, do this for each user (e.g., the correlation between ratings and predictions for user 5136 is 0.1698 -- only about 17%, and therefore not very good).  Then, apply the same correlation to the full matrices.  
6.1 report the top and bottom correlations (the bottom ones may be negative) by user (both user ID and correlation value).   
6.2 report the overall correlation between ratings and predictions.  


In [110]:
# correlations between user's ratings and predictions
corrs = {}

for u in df_ratings.index:
    corrs[u] = np.corrcoef(
        df_ratings.loc[u].dropna(), 
        df_preds.loc[u].dropna()
        )[0][1].round(3)
    
pd.Series(corrs).sort_values(ascending=False)

5261    0.921
442     0.774
2267    0.634
918     0.459
2492    0.414
2824    0.348
4942    0.348
5136    0.170
3556   -0.096
860    -0.148
dtype: float64

In [106]:
# overall correlations
np.corrcoef(
    pd.Series(df_ratings.values.flatten()).dropna(),
    pd.Series(df_preds.values.flatten()).dropna()
)[0][1]

np.float64(0.566447514804282)